Added method setMargin to btConvexHullShape and 3 methods to btCompou…

yomboprime opened the pull request:
…ndShape.

Added the method `setMargin` to `btConvexHullshape`, and three methods needed to destroy the sub-shapes of a `btCompoundShape` without having to remember them externally, since Bullet does not destroy them when you destroy the `btCompoundShape`.

The new methods in `btCompoundShape` are `removeChildShapeByIndex`, `getNumChildShapes` and `getChildShape`

All methods tested, and the tests passed.

kripken commented:
Nice!

Would it be possible to add tests for these things? If it's not too hard that would be good, to prevent regressions in the future. I can help integrating them in the test suite.

yomboprime commented:
Sure, but I don't know a lot of python and it seems complicated, but I'll try.
I don't know how to test the setMargin method.
The other methods of compound shape seems easy to test.

yomboprime commented:
I'm executing the tests with
`nodejs ammo.js test.js`
But shouldn't it print the `Ok.` Message? It outputs nothing.

kripken commented:
I can do the python parts. I just need a testcase in JS.

The problem with calling node like that is that it only runs the first file, and passes the second as an argument to it, so `node program.js 1 2 3` calls `program.js` with arguments `1,2,3`. You can either write a little node script to load multiple scripts, or just concatenate `ammo.js` and the testcase in a temp file, and run that single combined file.

yomboprime commented:
Well, I concatenated the files and now it says `ReferenceError: print is not defined`

yomboprime commented:
I've added a test file, `compoundShape.js`. It tests all methods I have added so far.
But I have to use `console.log('ok.')` Instead of `print`. Where is `print` defined?
I do `cat ammo.js testutils.js compoundShape.js > t.js` to test it and works ok.

yomboprime commented:
Justa added `convexHullShape` to the test and more `setMargin` calls

kripken commented:
`print` is present in most JS engines, but I think not node. Perhaps replacing that with `console.log` would work.

yomboprime commented:
Yeah, it works with console.log, it is already with that.

yomboprime commented:
Do you think the test `compoundShape.js` is all right?

kripken commented:
Would it also be possible to test some values using `assertEq` as well, so see that they can be read back properly? I'm not an expert on the bullet api though so I'm not sure if the changes here make sense to test that way.

yomboprime commented:
Yes, `getMargin` exists so I can read back and test it. I'll have to add it to all shapes, give me some minutes to do it.

yomboprime commented:
I have added the `getMargin` method to all shapes and used it in the tests.
I have let a small delta in the comparison though, since there are small differences (0.01 original versus 0.009999997... etc)

yomboprime commented:
Added a comment to clarify the sphere shape margin. And a missing `assertEq` test to the compound shape margin.

kripken commented:
Great, thanks! I'll merge and hook it up to the test runner.

yomboprime commented:
Thank you!

yomboprime commented:
BTW, note that currently it is using `console.log` instead of `print` to print the final message 'ok.'

kripken commented:
Yes, I noticed, but it works in other engines too - I think the test harness adds in console.log if needed, somewhere.

yomboprime commented:
All right then.
